test/crowdsale/ZxcCrowdsale.test.js

```diff
@@ -0,0 +1,583 @@
+      // Set crowdsale contract ZXC allowance
+      await token.approve(crowdsale.address, crowdSaleZxcCap, {from: tokenOwner});
+      await increaseTimeTo(startTimeSaleNoBonus + duration.seconds(30));
+
+      await assertRevert(crowdsale.buyTokens(buyerOne, {from: buyerOne, value: weiAmount}));
+    });
+
+    it('buyTokens should purchase tokens for minimum deposit amount', async () => {
+      let weiAmount = ether(1);
+      let expectedTokens = weiAmount.mul(rate);
+
+      // Set crowdsale contract ZXC allowance
+      await token.approve(crowdsale.address, crowdSaleZxcCap, {from: tokenOwner});
+      await increaseTimeTo(startTimeSaleNoBonus + duration.seconds(30));
+
+      await crowdsale.buyTokens(buyerOne, {from: buyerOne, value: weiAmount});
+      let actualTokens = await token.balanceOf(buyerOne);
+      assert.strictEqual(actualTokens.toString(), expectedTokens.toString());
+    });
+
+    it('buyTokens should purchase tokens if sold token amount hits crowdsale cap', async () => {
+      let weiAmount = ether(3);
+      let crowdsaleCap = ether(3).mul(rate);
+      let expectedTokens = weiAmount.mul(rate);
+
+      // Set crowdsale contract ZXC allowance
+      await token.approve(crowdsale.address, crowdsaleCap, {from: tokenOwner});
+      await increaseTimeTo(startTimeSaleNoBonus + duration.seconds(30));
+
+      await crowdsale.buyTokens(buyerOne, {from: buyerOne, value: weiAmount});
+      let actualTokens = await token.balanceOf(buyerOne);
+      assert.strictEqual(actualTokens.toString(), expectedTokens.toString());
+    });
+
+    it('buyTokens should fail purchasing if sold token amount goes over crowdsale cap', async () => {
+      let weiAmount = ether(3.1);
+      let crowdsaleCap = ether(3).mul(rate);
+
+      // Set crowdsale contract ZXC allowance
+      await token.approve(crowdsale.address, crowdsaleCap, {from: tokenOwner});
+      await increaseTimeTo(startTimeSaleNoBonus + duration.seconds(30));
+
+      await assertRevert(crowdsale.buyTokens(buyerOne, {from: buyerOne, value: weiAmount}));
+    });
+
+    it('buyTokens should fail purchasing tokens if transferFrom fails', async () => {
+      let weiAmount = ether(2.1);
+
+      await increaseTimeTo(startTimeSaleNoBonus + duration.seconds(30));
+
+      // Crowdsale ZXC allowance not set, transferFrom fails
+      await assertRevert(crowdsale.buyTokens(buyerOne, {from: buyerOne, value: weiAmount}));
+    });
+
+    it('fallback function should purchase tokens', async () => {
+      let weiAmount = ether(8.05113);
+      let expectedSoldTokens = weiAmount.mul(rate);
+      let startWalletBalance = await web3.eth.getBalance(wallet);
+
+      // Set crowdsale contract ZXC allowance
+      await token.approve(crowdsale.address, crowdSaleZxcCap, {from: tokenOwner});
+      await increaseTimeTo(startTimeSaleNoBonus + duration.seconds(30));
+
+      let { logs } = await crowdsale.buyTokens(buyerOne, {from: buyerOne, value: weiAmount});
```

Comment:
This is not testing the fall back function!

Reply:
Nice catch. Copy/paste typo :(

Reply:
f29e6ab